// express-badge/src/badgeFunctions.js
const PromiseFactory = require('logged-promise');
const RequestFactory = require('logged-request-promise');
const ModelFactory = require('sequelize-model-factory');
const _ = require('lodash');
const mapper = require('./mapper');
const asyncEach = require('async/each');

let config;

module.exports.init = (serviceInit) => {
  const Promise = PromiseFactory.init(serviceInit);
  const Request = RequestFactory.init(serviceInit);
  const DbModel = ModelFactory.init(serviceInit).get('DisplayToken');
  const { currentUser } = serviceInit;

  const badgeFunctions = {};

  badgeFunctions.validate = (_config) => {
    config = _config;

    if (_.isNil(serviceInit)) throw new Error('expres-badge validate: invalid request');
    if (_.isNil(config)) throw new Error('expres-badge validate: config is null or undefined');
    if (_.isNil(config.endpoints.userServiceUrl)) throw new Error('expres-badge validate: invalid user service endpoint');
    if (_.isNil(config.endpoints.roleServiceUrl)) throw new Error('expres-badge validate: invalid role service endpoint');
    if (_.isNil(config.endpoints.departmentServiceUrl)) throw new Error('expres-badge validate: invalid department service endpoint');
  };

  badgeFunctions.mapRequestParameters = (requestBody, requestParam) => {
    const badgeFields = {
      User: {},
      Role: {},
      Department: {},
    };

    // request body i döner.
    _.forIn(requestBody, (value, key) => {
      if (key.toLowerCase().indexOf('userid') > -1 && value) {
        if (Array.isArray(value)) {
          _.forEach(value, (item) => {
            badgeFields.User[item] = key;
          });
        } else { badgeFields.User[value] = key; }
      }

      if (key.toLowerCase().indexOf('roleid') > -1 && value) {
        if (Array.isArray(value)) {
          _.forEach(value, (item) => {
            badgeFields.Role[item] = key;
          });
        } else { badgeFields.Role[value] = key; }
      }

      if (key.toLowerCase().indexOf('departmentid') > -1 && value) {
        if (Array.isArray(value)) {
          _.forEach(value, (item) => {
            badgeFields.Department[item] = key;
          });
        } else { badgeFields.Department[value] = key; }
      }
    });

    // requestParamsı döner
    _.forIn(requestParam, (value, key) => {
      if (_.replace(key, '_', '').toLowerCase().indexOf('userid') > -1 && value) { badgeFields.User[value] = key; }

      if (_.replace(key, '_', '').toLowerCase().indexOf('roleid') > -1 && value) { badgeFields.Role[value] = key; }

      if (_.replace(key, '_', '').toLowerCase().indexOf('departmentid') > -1 && value) { badgeFields.Department[value] = key; }
    });

    badgeFields.User[currentUser.userId] = 'currentUser';

    return badgeFields;
  };

  badgeFunctions.createBadges = (badges) => {
    const saveBadge = (badgeItem, asyncCallback) => {
      asyncEach(Object.keys(badges[badgeItem]), (item, cb) => {
        badgeFunctions.createBadge(item, badgeItem)
          .then(() => cb(null))
          .catch(cb);
      }, itemErr => asyncCallback(itemErr));
    };

    return new Promise(badges, (res, rej) => {
      asyncEach(Object.keys(badges), saveBadge, (badgeErr) => {
        if (badgeErr) {
          return rej(badgeErr);
        }

        return res(null);
      });
    });
  };

  badgeFunctions.getBadge = (id, _subjectType) => {
    let options = {};

    if (_subjectType === 'User') {
      options = mapper.mapRequest(config, 'User', id, currentUser);
    } else {
      options = mapper.mapRequest(config, _subjectType, id, currentUser);
    }

    return Request(options).then(requestReponse =>
      DbModel.create(mapper.map(requestReponse.data, _subjectType)));
  };

  badgeFunctions.createBadge = (id, _subjectType) => new Promise(id, (res, rej) => {
    const options = {
      where: {
        subjectType: _subjectType,
        subjectId: id,
      },
    };

    DbModel.findOne(options).then((instance) => {
      if (!instance) {
        if (_subjectType === 'User' && id === currentUser.userId.toString()) {
          return DbModel.create(mapper.map(currentUser, 'CurrentUser'));
        }
        return badgeFunctions.getBadge(id, _subjectType);
      }
    }).then(res).catch(rej);
  });

  return badgeFunctions;
};
