// express-badge/src/mapper.js
const BasicUrl = require('basic-url');

const Mapper = {};

Mapper.map = (data, subjectType) => {
  let result = {};

  switch (subjectType) {
    case 'CurrentUser': result = Mapper.mapCurrentUserData(data); break;

    case 'User': result = Mapper.mapUserData(data); break;

    case 'Role': result = Mapper.mapData(data, subjectType); break;

    case 'Department': result = Mapper.mapData(data, subjectType); break;

    default: result = Mapper.mapUserData(data); break;
  }

  return result;
};

Mapper.mapCurrentUserData = (currentUser) => {
  const insertedData = {
    displayName1: currentUser.firstName,
    displayName2: currentUser.lastName,
    image: currentUser.avatar,
    subjectId: currentUser.userId,
    subjectType: 'User',
  };

  return insertedData;
};

Mapper.mapUserData = (dataUser) => {
  const insertedData = {
    displayName1: dataUser.firstName,
    displayName2: dataUser.lastName,
    image: dataUser.avatar,
    subjectId: dataUser.id,
    subjectType: 'User',
  };

  return insertedData;
};

Mapper.mapData = (dataRole, subjectType) => {
  const insertedData = {
    displayName1: dataRole.name,
    subjectId: dataRole.id,
    subjectType,
  };

  return insertedData;
};

Mapper.mapRequest = (config, subjectType, subjectId, currentUser) => {
  let url;
  let key;
  switch (subjectType) {
    case 'User':
      url = config.endpoints.userServiceUrl;
      key = 'user_id';
      break;

    case 'Role':
      url = config.endpoints.roleServiceUrl;
      key = 'role_id';
      break;
    case 'Department':
      url = config.endpoints.departmentServiceUrl;
      key = 'department_id';
      break;

    default:
      url = config.endpoints.userServiceUrl;
      key = 'user_id';
      break;
  }

  const UrlOptions = {
    url,
    params: {
      organization_id: currentUser.organizationId,
      [key]: subjectId,
    },
  };

  const options = {
    uri: BasicUrl(UrlOptions),
    method: 'get',
    isAuthenticated: true,
    json: true,
  };

  return options;
};

module.exports = Mapper;
