// express-badge/src/router.js
const ModelFactory = require('sequelize-model-factory');
const responser = require('express-responser');
const _ = require('lodash');
const Errors = require('./errors');
const Authenticator = require('express-authenticator');

module.exports.addRoute = (app, config) => {
  app.put('/client/displaytokens/:subject_type/:subject_id', [Authenticator(config).verifyClient], (req, res, next) => {
    req.check({
      displayName1: { notEmpty: true },
      subject_type: { notEmpty: true },
      subject_id: { notEmpty: true },
    });

    if (req.validationErrors()) {
      return responser.withError(next, new Errors.ValidationError(req.validationErrors(true)));
    }

    const where = {
      subjectType: req.params.subject_type,
      subjectId: req.params.subject_id,
    };

    const data = {
      displayName1: req.body.displayName1,
    };

    if (req.body.displayName2) { data.displayName2 = req.body.displayName2; }

    if (req.body.image) { data.image = req.body.image; }

    const DbModel = ModelFactory.init(req.serviceInit).get('DisplayToken');

    DbModel.findOne({ where }).then((instance) => {
      if (instance) { return DbModel.update(_.assignIn(instance, data)); }
      return null;
    }).then(updateResult => responser.success(res, updateResult))
      .catch(() => responser.withError(next, new Errors.DisplayTokenUpdate()));
  });
};
