// sequelize-model-factory/index.js
let flow = require('lodash/flow');
const Sequelize = require('sequelize');
const fs = require('fs');
const { isNil, isFunction, has, isEmpty, get } = require('lodash');

const PromiseFactory = require('logged-promise');
//const config = require('../config.js');
//const logger = require('../helpers/logger').logger;
const Op = Sequelize.Op;
const operatorsAliases = {
  $eq: Op.eq,
  $ne: Op.ne,
  $gte: Op.gte,
  $gt: Op.gt,
  $lte: Op.lte,
  $lt: Op.lt,
  $not: Op.not,
  $in: Op.in,
  $notIn: Op.notIn,
  $is: Op.is,
  $like: Op.like,
  $notLike: Op.notLike,
  $iLike: Op.iLike,
  $notILike: Op.notILike,
  $regexp: Op.regexp,
  $notRegexp: Op.notRegexp,
  $iRegexp: Op.iRegexp,
  $notIRegexp: Op.notIRegexp,
  $between: Op.between,
  $notBetween: Op.notBetween,
  $overlap: Op.overlap,
  $contains: Op.contains,
  $contained: Op.contained,
  $adjacent: Op.adjacent,
  $strictLeft: Op.strictLeft,
  $strictRight: Op.strictRight,
  $noExtendRight: Op.noExtendRight,
  $noExtendLeft: Op.noExtendLeft,
  $and: Op.and,
  $or: Op.or,
  $any: Op.any,
  $all: Op.all,
  $values: Op.values,
  $col: Op.col
};

const db = {};

function getFlows(serviceInit, entityName) {

  function optionSetAttributes(options) {
    options = options || {};
    options.opt = options.opt || {};

    if (options.opt.attributes && isFunction(options.opt.attributes)) {
      options.opt.attributes = options.opt.attributes(db.sequelize.fn, db.sequelize.col);
    }

    return options;
  }

  function optionSetIncludeModel(options) {
    options = options || {};
    options.opt = options.opt || {};

    if (options.opt.include && options.opt.include.length > 0) {
      options.opt.include.forEach((value) => {
        if (!value || !value.modelName) {
          throw new Error('optionSetIncludeModel: options.opt.include[].modelName is null or undefined');
        }

        const organizationId = serviceInit.getOrganizationId(true);
        value.model = db[value.modelName];
        if (has(db, `${value.modelName}.rawAttributes.organizationId`) && organizationId && get(db, `${entityName}.associations.${value.as}.options.setOrganizationIdForInclude`, true)) {
          if (!has(value, 'required') && isEmpty(value.where)) {
            value.required = false;
          }
          value.where = value.where || {};

          //FullStay Master-Fiyat kısmında OrganizationId'nin null setlenebilmesi için 
          //Modelde allowNull ve allowNullPrimary kontrol edilerek Or ekleniyor
          if (
            value.model.rawAttributes.organizationId
            && value.model.rawAttributes.organizationId.allowNull
            && value.allowNullPrimary
          ) {
            value.where[Op.or] =
              [
                { organizationId: null },
                { organizationId: organizationId }
              ];
          } else {
            value.where.organizationId = organizationId;
          }

        }

        if (value.include && value.include.length > 0) {
          value.include.forEach((value2) => {
            if (!value2 || !value2.modelName) {
              throw new Error('optionSetIncludeModel: options.opt.include[].modelName is null or undefined');
            }
            value2.model = db[value2.modelName];
            if (has(db, `${value2.modelName}.rawAttributes.organizationId`) && organizationId && get(db, `${value2.modelName}.associations.${value2.as}.options.setOrganizationIdForInclude`, true)) {
              if (!has(value2, 'required') && isEmpty(value2.where)) {
                value2.required = false;
              }
              value2.where = value2.where || {};

              //FullStay Master-Fiyat kısmında OrganizationId'nin null setlenebilmesi için 
              //Modelde allowNull ve allowNullPrimary kontrol edilerek Or ekleniyor
              if (
                value2.model.rawAttributes.organizationId
                && value2.model.rawAttributes.organizationId.allowNull
                && value2.allowNullPrimary
              ) {
                value2.where[Op.or] =
                  [
                    { organizationId: null },
                    { organizationId: organizationId }
                  ];
              } else {
                value2.where.organizationId = organizationId;
              }
            }
          });
        }
      });
    }

    return options;
  }

  function optionSetOrderModel(options) {
    options = options || {};
    options.opt = options.opt || {};

    if (options.opt.order && options.opt.order.length > 0) {
      options.opt.order.forEach((value) => {
        if (!value[0] || !value[1]) {
          throw new Error('optionSetOrderModel: options.opt.order[] is null or undefined');
        }
        if (value[0].modelName === entityName) {
          value.splice(0, 1);
        }

        if (db[value[0].modelName]) {
          value[0] = { model: db[value[0].modelName], as: value[0].as };
          value[1] = db.sequelize.col(value[1]);
        } else {
          value[0] = db.sequelize.col(value[0]);
        }
      });
    }

    return options;
  }

  function optionSetIdForOrganization(options) {
    options = options || {};
    options.opt = options.opt || {};
    // eger organization cekilen entity ise ID columnindan setlensin
    let organizationId = serviceInit.getOrganizationId(true);

    if (entityName === 'Organization' && organizationId) {
      options.opt.where = options.opt.where || {};
      if (organizationId === '*') {
        options.opt.where.id = { $ne: null };
      } else {
        options.opt.where.id = organizationId;
      }
    }

    return options;
  }

  function optionSetGranteeOrganizationId(options) {
    options = options || {};
    options.opt = options.opt || {};

    if (
      db[entityName].rawAttributes.hasOwnProperty('granteeOrganizationId')
      && serviceInit.crossOptions
      && serviceInit.crossOptions.cross
      && serviceInit.currentUser
      && serviceInit.currentUser.crossOrganization
      && serviceInit.currentUser.crossOrganization.granteeOrganizationId
    ) {

      options.opt.where = options.opt.where || {};
      options.opt.where.granteeOrganizationId = serviceInit.currentUser.crossOrganization.granteeOrganizationId;
    }

    return options;
  }

  function optionSetOrganizationId(options) {
    options = options || {};
    options.opt = options.opt || {};
    //eger organization id modelde varsa setlensin
    let organizationId = serviceInit.getOrganizationId(true);
    if (db[entityName].rawAttributes.hasOwnProperty('organizationId') && organizationId) {
      options.opt.where = options.opt.where || {};
      if (organizationId === '*') {
        options.opt.where.organizationId = { $ne: null };
      } else {

        //FullStay Master-Fiyat kısmında OrganizationId'nin null setlenebilmesi için 
        //Modelde allowNull ve allowNullPrimary kontrol edilerek Or ekleniyor
        if (
          db[entityName].rawAttributes.organizationId
          && db[entityName].rawAttributes.organizationId.allowNull
          && options.opt.allowNullPrimary
        ) {
          options.opt.where[Op.or] =
            [
              { organizationId: null },
              { organizationId: organizationId }
            ];
        } else {
          options.opt.where.organizationId = organizationId;
        }
      }
    }

    return options;
  }

  function optionSetOrganizationIdForUpdateOrDelete(options) {
    options = options || {};
    options.opt = options.opt || {};
    //eger organization id modelde varsa setlensin
    let organizationId = serviceInit.getOrganizationId(false);

    if (db[entityName].rawAttributes.hasOwnProperty('organizationId') && organizationId) {
      options.opt.where = options.opt.where || {};
      options.opt.where.organizationId = organizationId;

    }

    return options;
  }

  function optionSetPrimaryKeyForUpdate(options) {
    if (isNil(options.data))
      throw new Error('optionSetPrimaryKeyForUpdate: options.data is null or undefined');
    //eger opt yoksa opt yi kendimiz yaratacagiz
    if (!options.opt || !options.opt.where) {
      //options.opt = { where: {} };
      options.opt = options.opt || {};
      options.opt.where = {};
      // eger bir opt yoksa ve gelen update datasinda PK varsa opt.where e set et
      if (options.data.hasOwnProperty(db[entityName].primaryKeyField))
        options.opt.where[db[entityName].primaryKeyField] = options.data[db[entityName].primaryKeyField];
    }

    if (!options.opt.where || !options.opt.where[db[entityName].primaryKeyField])
      throw new Error('optionSetPrimaryKeyForUpdate: options.opt.where is null or undefined or does not have primary key');

    return options;
  }

  function dataTrimForUpdate(options) {
    if (isNil(options.data))
      throw new Error('dataTrimForUpdate: options.data is null or undefined');

    let trimmedUpdateData = {};
    Object.keys(options.data).forEach((value) => {

      if (options.data[value] === null) {
        trimmedUpdateData[value] = null;
      } else if (typeof options.data[value] !== "undefined")
        trimmedUpdateData[value] = options.data[value];
    });
    options.data = trimmedUpdateData;
    return options;
  }

  function dataSetOrganizationIdForCreate(options) {
    let organizationId = serviceInit.getOrganizationId(false);
    if (db[entityName].rawAttributes.hasOwnProperty('organizationId') && organizationId) {
      options.data.organizationId = organizationId;
    }
    return options;
  }

  function dataSetAuditFieldsForCreate(options) {

    if (db[entityName].rawAttributes.hasOwnProperty('createdBy')) {

      if (serviceInit.currentUser && serviceInit.currentUser.delegateUser)
        options.data.createdBy = serviceInit.currentUser.delegateUser.username;
      else if (serviceInit.currentUser)
        options.data.createdBy = serviceInit.currentUser.username;
      else if (serviceInit.currentClient)
        options.data.createdBy = serviceInit.currentClient.clientName;
    }

    if (db[entityName].rawAttributes.hasOwnProperty('updatedUserId') && serviceInit.currentUser) {
      options.data.updatedUserId = serviceInit.currentUser.userId;
    }

    if (db[entityName].rawAttributes.hasOwnProperty('logRequestId') && serviceInit.requestId) {
      options.data.logRequestId = serviceInit.requestId;
    }

    return options;
  }

  function dataSetAuditFieldsForUpdate(options) {
    if (db[entityName].rawAttributes.hasOwnProperty('updated_at'))
      options.data.updatedAt = new Date();

    if (db[entityName].rawAttributes.hasOwnProperty('updatedBy')) {

      if (serviceInit.currentUser && serviceInit.currentUser.delegateUser)
        options.data.updatedBy = serviceInit.currentUser.delegateUser.username;
      else if (serviceInit.currentUser)
        options.data.updatedBy = serviceInit.currentUser.username;
      else if (serviceInit.currentClient)
        options.data.updatedBy = serviceInit.currentClient.clientName;
    }

    if (db[entityName].rawAttributes.hasOwnProperty('updatedUserId') && serviceInit.currentUser) {
      options.data.updatedUserId = serviceInit.currentUser.userId;
    }

    if (db[entityName].rawAttributes.hasOwnProperty('logRequestId') && serviceInit.requestId) {
      options.data.logRequestId = serviceInit.requestId;
    }

    return options;
  }

  function dataSetOrganizationIdForBulkCreate(options) {

    let organizationId = serviceInit.getOrganizationId(false);

    if (db[entityName].rawAttributes.hasOwnProperty('organizationId') && organizationId) {

      options.data.forEach((item) => {
        item.organizationId = organizationId;
      });
    }
    return options;
  }

  function dataSetAuditFieldsForBulkCreate(options) {

    if (db[entityName].rawAttributes.hasOwnProperty('createdBy')) {

      let createdBy;

      if (serviceInit.currentUser && serviceInit.currentUser.delegateUser)
        createdBy = serviceInit.currentUser.delegateUser.username;
      else if (serviceInit.currentUser)
        createdBy = serviceInit.currentUser.username;
      else if (serviceInit.currentClient)
        createdBy = serviceInit.currentClient.clientName;

      options.data.forEach((item) => {
        item.createdBy = createdBy;
      });
    }

    return options;
  }

  return {
    selectFlow: flow(
      optionSetOrganizationId,
      optionSetIdForOrganization,
      optionSetIncludeModel,
      optionSetOrderModel,
      optionSetGranteeOrganizationId,
      optionSetAttributes
    ),
    destroyFlow: flow(
      optionSetOrganizationIdForUpdateOrDelete,
      optionSetIdForOrganization,
      optionSetGranteeOrganizationId
    ),
    updateFlow: flow(
      dataTrimForUpdate,
      optionSetPrimaryKeyForUpdate,
      dataSetAuditFieldsForUpdate,
      optionSetOrganizationIdForUpdateOrDelete,
      optionSetIdForOrganization,
      optionSetGranteeOrganizationId
    ),
    createFlow: flow(
      dataSetOrganizationIdForCreate,
      dataSetAuditFieldsForCreate
    ),
    bulkCreateFlow: flow(
      dataSetOrganizationIdForBulkCreate,
      dataSetAuditFieldsForBulkCreate
    )
  };
}

const mapObject = (results) => {
  const response = results.dataValues;

  if (response) {
    Object.keys(response).forEach((key) => {
      if (Array.isArray(response[key])) {
        response[key] = response[key].map(v => v.dataValues ? mapObject(v) : v);
      } else if (response[key] && response[key].dataValues) {
        response[key] = mapObject(response[key]);
      }
    });
  }
  return response;
};

const mappersFactory = (serviceInit) => {
  return {
    selectMapper: (results) => {
      if (isNil(results)) {
        return null;
      }

      if (Array.isArray(results)) {
        const mappedList = results.map(val => mapObject(val));
        return !mappedList || !mappedList.length || mappedList.length === 0 ? null : mappedList;
      }

      return mapObject(results);
    },
    createMapper: (result) => {
      if (result && result.Model && result.Model.tableName === 'organizations') {
        if (serviceInit.currentUser) {
          serviceInit.currentUser.organizationId = result.dataValues.id;
        }
        if (serviceInit.currentClient) {
          serviceInit.currentClient.organizationId = result.dataValues.id;
        }
      }
      return mapObject(result);
    },
    bulkCreateMapper: (results) => {

      let response = [];

      if (Array.isArray(results)) {
        response = results.map(val => mapObject(val));
      } else {
        response = mapObject(results);
      }

      return response;
    },
    updateMapper: (updateResult, factoryOptions) => {
      return {
        affectedRowCount: updateResult[0],
        instance: factoryOptions.data
      };
    },
    decrementMapper: (updateResult, factoryOptions) => {
      return {
        affectedRowCount: updateResult[0][1],
        instance: factoryOptions.data
      };
    },
    destroyMapper: (affectedRowCount) => {
      return { affectedRowCount: affectedRowCount };
    },
    countMapper: (rowCount) => {
      return { rowCount: rowCount };
    },
    findAndCountAllMapper: (results) => {

      if (Array.isArray(results.rows)) {
        results.rows = results.rows.map(val => mapObject(val));
      }

      return results;
    },
  }
};

module.exports.appStart = (logger, config, onSequelizeInit) => {
  validate(logger, config);
  const sequelize = new Sequelize(config.database.database, config.database.user, config.database.password, {
    host: config.database.host,
    port: config.database.port,
    dialect: config.database.dialect,
    storage: config.database.storage,
    operatorsAliases,
    pool: {
      max: config.database.maxConnection || 60,
      min: 0,
      idle: 10000
    },
    logging: (errorMessage) => {
      logger.debug(errorMessage);
    }
  });

  fs.readdir(config.models, (err, files) => {
    if (isNil(files)) {
      return false;
    }

    files.forEach((file) => {
      let fileName = file.split('.')[0];
      db[fileName] = sequelize.import(config.models + '/' + file);
    });
    if (isFunction(onSequelizeInit)) {
      onSequelizeInit(db);
    }
  });

  db.sequelize = sequelize;
  db.constants = {
    ISOLATION_LEVELS: Sequelize.Transaction.ISOLATION_LEVELS,
    TYPES: Sequelize.Transaction.TYPES,
    LOCK: Sequelize.Transaction.LOCK
  };
  db.customFunctions = {
    literal: sequelize.literal
  }
};
module.exports.Op = Sequelize.Op;
module.exports.init = (serviceInit, opt) => {
  if (!serviceInit || !serviceInit.requestId) throw new Error('Invalid serviceInit in sequelize-model-factory');
  return {
    get: (entityName) => {
      let Promise = PromiseFactory.init(serviceInit);
      let flows = getFlows(serviceInit, entityName);
      const mappers = mappersFactory(serviceInit);
      let run = (factoryOptions, callback) => {
        let loggedData = (opt && opt.hideLogResponse) || get(factoryOptions, 'opt.hideLogResponse', false) ? null : { data: factoryOptions.data, opt: factoryOptions.opt };
        return new Promise(loggedData, function (res, rej) {
          callback(factoryOptions.flow(factoryOptions))
            .then((result) => res(factoryOptions.mapper(result, factoryOptions)))
            .catch(rej);
        }, opt || {});
      };

      return {
        customFunctions: db.customFunctions,
        constants: db.constants,
        transaction: (opt) => db.sequelize.transaction(opt),
        findOne: (opt) => {
          if (isNil(opt))
            throw new Error('sequelize-model-factory findOne: id is null or undefined');
          let factoryOptions = {
            mapper: mappers.selectMapper,
            flow: flows.selectFlow,
            opt: opt
          };
          return run(factoryOptions, (composed) => db[entityName].findOne(composed.opt));
        },
        findById: (id, opt) => {
          if (isNil(id))
            throw new Error('sequelize-model-factory findById: id is null or undefined');

          opt = opt || {};
          opt.where = {};

          opt.where[db[entityName].primaryKeyField] = id;

          let factoryOptions = {
            mapper: mappers.selectMapper,
            flow: flows.selectFlow,
            opt: opt
          };
          return run(factoryOptions, (composed) => db[entityName].findOne(composed.opt));
        },
        findAll: (opt) => {
          let factoryOptions = {
            mapper: mappers.selectMapper,
            flow: flows.selectFlow,
            opt: opt
          };
          return run(factoryOptions, (composed) => db[entityName].findAll(composed.opt));
        },
        findAndCountAll: (opt) => {
          let factoryOptions = {
            mapper: mappers.findAndCountAllMapper,
            flow: flows.selectFlow,
            opt: opt
          };

          return run(factoryOptions, (composed) => db[entityName].findAndCountAll(composed.opt));
        },
        update: (updateData, opt) => {

          if (isNil(updateData))
            throw new Error('sequelize-model-factory update: update data is null or undefined');

          let factoryOptions = {
            mapper: mappers.updateMapper,
            flow: flows.updateFlow,
            opt: opt,
            data: updateData
          };

          return run(factoryOptions, (composed) => db[entityName].update(composed.data, composed.opt));
        },
        destroy: (opt) => {

          if (isNil(opt))
            throw new Error('sequelize-model-factory destroy: opt is null or undefined');

          const factoryOptions = {
            mapper: mappers.destroyMapper,
            flow: flows.destroyFlow,
            opt: opt
          };

          if (db[entityName].rawAttributes.hasOwnProperty('isDeleted')) {
            factoryOptions.data = {
              isDeleted: db.sequelize.col('id'),
              deleted_at: new Date()
            };
            return run(factoryOptions, (composed) => db[entityName].update(composed.data, composed.opt));
          } else {
            return run(factoryOptions, (composed) => db[entityName].destroy(composed.opt));
          }
        },
        create: (createData, opt) => {
          if (isNil(createData))
            throw new Error('sequelize-model-factory createData: create data is null or undefined');

          let factoryOptions = {
            mapper: mappers.createMapper,
            flow: flows.createFlow,
            opt: opt,
            data: createData
          };
          return run(factoryOptions, (composed) => db[entityName].create(composed.data, composed.opt));
        },
        count: (opt) => {
          if (isNil(opt))
            throw new Error('sequelize-model-factory count: opt is null or undefined');

          let factoryOptions = {
            mapper: mappers.countMapper,
            flow: flows.selectFlow,
            opt: opt
          };
          return run(factoryOptions, (composed) => db[entityName].count(composed.opt));
        },
        decrement: (values, opt) => {
          if (isNil(values)) {
            throw new Error('sequelize-model-factory count: values is null or undefined');
          }

          if (isNil(opt)) {
            throw new Error('sequelize-model-factory count: opt is null or undefined');
          }

          let factoryOptions = {
            mapper: mappers.decrementMapper,
            flow: flows.updateFlow,
            opt: opt,
            data: values
          };
          return run(factoryOptions, (composed) => {
            delete composed.data.updatedAt;
            return db[entityName].decrement(composed.data, composed.opt)
          });
        },
        bulkCreate: (createData, opt) => {

          if (isNil(createData))
            throw new Error('sequelize-model-factory bulkCreate: create data is null or undefined');

          if (!Array.isArray(createData))
            throw new Error('sequelize-model-factory bulkCreate: create data is not array');

          let factoryOptions = {
            mapper: mappers.bulkCreateMapper,
            flow: flows.bulkCreateFlow,
            opt: opt,
            data: createData
          };
          return run(factoryOptions, (composed) => db[entityName].bulkCreate(composed.data, composed.opt));
        }
      };
    }
  };
};

function validate(logger, config) {
  if (isNil(logger)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: logger is null or undefined');
  if (isNil(config)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config is null or undefined');
  if (isNil(config.database)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database is null or undefined');
  if (isNil(config.database.database)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database.database is null or undefined');
  if (isNil(config.database.user)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database.user is null or undefined');
  if (isNil(config.database.password)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database.password is null or undefined');
  if (isNil(config.database.host)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database.host is null or undefined');
  if (isNil(config.database.port)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database.port is null or undefined');
  if (isNil(config.database.dialect)) throw new Error('sequelize-model-factory validate: Invalid appStart parameters: config.database.dialect is null or undefined');
}
